### PR TITLE
[No QA] Include newest version of eslint-config-expensify with new rule of no-multiple-onyx-in-file

### PR DESCRIPTION
**Pull request by studentofcoding**
### Details

**ESLint Rules extend - Create an ESLint rule to detect multiple uses of withOnyx**

This change only affects local development and does not need to be QAed and Affected files (MoneyRequestConfirmPage.js, MoneyRequestTagPage.js, EditRequestPage.js, and OptionRowLHNData.js) would have  `// eslint-disable-next-line` right now & would be fixed separately on Another issue after the PR is merged, based on our last discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1695202740378369?thread_ts=1695197339.544539&cid=C01GTK53T8Q)

### Fixed Issues

$ [#27463](https://github.com/Expensify/App/issues/27463)
PROPOSAL: https://github.com/Expensify/App/issues/27463#issuecomment-1720088085

### Tests

1. Check that we have the correct version of "eslint-config-expensify": "^2.0.39", and re-install the Project (better if we do a clean install)
2. Search `withOnyx({` on your code editor and click project with more than 1 `withOnyx({` like on `src/components/LHNOptionsList/OptionRowLHNData.js` below https://github.com/Expensify/App/blob/70e9fa0a77aba91abd12d8d0023ce10e7bfe80a7/src/components/LHNOptionsList/OptionRowLHNData.js#L182-L217
3. We will see the error appearing as it has more than 1 `withOnyx({`
Screenshot 2023-09-20 at 15 26 13
4. Also we can check the rule on another file that only have 1 `withOnyx({` and add another `withOnyx({` with random details and see that the error appear
Screenshot 2023-09-20 at 15 28 21

- [x] Verify that no errors appear in the JS console

### QA Steps

This change only affects local development and does not need to be QAed

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

cc: @neil-marcellini @mollfpr @tgolen

### CODE REVIEW

**Comment by mollfpr** [comment]
Screenshot 2023-09-21 at 12 22 19


@studentofcoding I think you put the issue link wrong; that's why Melvin doesn't pull the correct engineer.

**Comment by studentofcoding** [comment]
> 27463

Hmm, It's included on the right Issue I believe @mollfpr ?
Screenshot 2023-09-21 at 11 24 42
Screenshot 2023-09-21 at 11 24 27

**Comment by mollfpr** [comment]
Yes, the link is correct, but you paste it wrong. You should only paste the link https://github.com/Expensify/App/issues/27463, not [27463](https://github.com/Expensify/App/issues/27463).

**Comment by studentofcoding** [comment]
> Yes, the link is correct, but you paste it wrong. You should only paste the link #27463, not [27463](https://github.com/Expensify/App/issues/27463).

Noted, It's edited, do I need to do anything else @mollfpr ?

**Comment by mollfpr** [comment]
@studentofcoding Nope, it looks good now! Just make sure you doing it right on the next PR 👍

**Comment by studentofcoding** [comment]
> @studentofcoding Nope, it looks good now! Just make sure you doing it right on the next PR 👍

Yess, will do @mollfpr, thanks for pointing it out!

File: src/pages/iou/MoneyRequestTagPage.js [hunk 1]
```diff
@@ -109,6 +109,14 @@ export default compose(
             key: ONYXKEYS.IOU,
         },
     }),
+    //  eslint-disable-next-line
+    withOnyx({
```
Comment (tgolen) [inline src/pages/iou/MoneyRequestTagPage.js:113]: I don't think this PR should be adding any new uses of `withOnyx`. Was it a mistake to include this?

File: src/pages/iou/MoneyRequestTagPage.js [hunk 2]
```diff
@@ -109,6 +109,14 @@ export default compose(
             key: ONYXKEYS.IOU,
         },
     }),
+    //  eslint-disable-next-line
```
Comment (tgolen) [inline src/pages/iou/MoneyRequestTagPage.js:112]: Please update all of these to use the format for only disabling the specific rule `no-multiple-onyx-in-file`

File: src/pages/iou/MoneyRequestTagPage.js [hunk 3]
```diff
@@ -109,6 +109,14 @@ export default compose(
...
+    withOnyx({
+        report: {
+            // Fetch report ID from IOU participants if no report ID is set in route
+            key: ({route, iou}) => `${ONYXKEYS.COLLECTION.REPORT}${lodashGet(route, 'params.reportID', '') || lodashGet(iou, 'participants.0.reportID', '')}`,
+        },
+    }),
```
Comment (mollfpr) [inline src/pages/iou/MoneyRequestTagPage.js:118]: @studentofcoding This change is not necessary, right? It was removed here https://github.com/Expensify/App/commit/6a47b41dbfb39ed6cd46ba0493aaca7a00888ac6

**Review by tgolen — CHANGES_REQUESTED** [review]

**Comment by studentofcoding** [comment]
Hey @cristipaval, This PR is a follow up PR for new eslint-config-expensiy, looking forward to your approval!

**Comment by mollfpr** [comment]
@studentofcoding Neil Is the one assigned to the issue. Also, the internal engineer will review after C+ approves the PR. So you can wait till that time before ping them again.

File: src/pages/iou/MoneyRequestTagPage.js (hunk 3 above)
Comment (studentofcoding) [reply src/pages/iou/MoneyRequestTagPage.js:118]: Ahh yes, I missed it and added it again

File: src/pages/iou/MoneyRequestTagPage.js (hunk 1 above)
Comment (studentofcoding) [reply src/pages/iou/MoneyRequestTagPage.js:113]: Yess, this was a leftover that I added again, it's deleted now @tgolen

File: src/pages/iou/MoneyRequestTagPage.js (hunk 2 above)
Comment (studentofcoding) [reply src/pages/iou/MoneyRequestTagPage.js]: Noted, It's already updated to // eslint-disable-next-line rulesdir/no-multiple-onyx-in-file

**Comment by studentofcoding** [comment]
> @studentofcoding Neil Is the one assigned to the issue. Also, the internal engineer will review after C+ approves the PR. So you can wait till that time before ping them again.

Noted thanks!

**Review by tgolen — DISMISSED** [review]
Looks great, thank you!

**Comment by studentofcoding** [comment]
> Looks great, thank you!

Thank you too!

File: .eslintrc.js [hunk 4]
```diff
@@ -39,6 +39,7 @@ module.exports = {
             files: ['*.js', '*.jsx', '*.ts', '*.tsx'],
             rules: {
                 'rulesdir/onyx-props-must-have-default': 'off',
+                'rulesdir/no-multiple-onyx-in-file': 'error',
```
Comment (mollfpr) [inline .eslintrc.js:42]: This is not necessary. We already define the rule in `eslint-config-expensify/rules/expensify`.

Comment (studentofcoding) [reply .eslintrc.js]: Noted, I've checked after remove it and It's working as expected

**Comment by studentofcoding** [comment]
Please, kindly check, re-approve again @tgolen @mollfpr

**Comment by mollfpr** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by mollfpr — APPROVED** [review]
LGTM and tests well 👍

**Comment by studentofcoding** [comment]
> LGTM and tests well 👍

Thanks!

**Review by cristipaval — APPROVED** [review]

**Comment by cristipaval** [comment]
Merging, as @tgolen already approved this.

**Pull request merged**